cupyx/distributed/array/_array.py

```diff
@@ -0,0 +1,424 @@
+from itertools import chain
+from typing import Any, Callable, Iterable, Optional, Union
+
+import numpy
+from numpy.typing import ArrayLike
+from numpy.typing import DTypeLike
+
+import cupy
+from cupy._core.core import ndarray
+import cupy._creation.from_data as _creation_from_data
+from cupy.cuda.device import Device
+from cupy.cuda.stream import Stream
+from cupy.cuda.stream import get_current_stream
+
+from cupyx.distributed.array import _chunk
+from cupyx.distributed.array._chunk import _Chunk
+from cupyx.distributed.array import _data_transfer
+from cupyx.distributed.array._data_transfer import _Communicator
+from cupyx.distributed.array import _elementwise
+from cupyx.distributed.array import _index_arith
+from cupyx.distributed.array import _modes
+from cupyx.distributed.array import _reduction
+from cupyx.distributed.array import _linalg
+
+
+class _MultiDeviceDummyMemory(cupy.cuda.memory.Memory):
+    pass
+
+
+class _MultiDeviceDummyPointer(cupy.cuda.memory.MemoryPointer):
+    @property
+    def device(self) -> Device:
+        # This override is needed to assign an invalid device id
+        # Since the array is not residing in a single device now
+        return Device(-1)
+
+
+class DistributedArray(ndarray):
+    """Multi-dimensional array distributed across multiple CUDA devices.
+
+    This class implements some elementary operations that :class:`ndarray`
+    provides. The array content is split into chunks, contiguous arrays
+    corresponding to slices of the original array. Note that one device can
+    hold multiple chunks.
+    """
+
+    _chunks_map: dict[int, list[_Chunk]]
+    _streams: dict[int, Stream]
+    _mode: _modes._Mode
+    _comms: dict[int, _Communicator]
+
+    def __new__(
+        cls, shape: tuple[int, ...], dtype: DTypeLike,
+        chunks_map: dict[int, list[_Chunk]],
+        mode: Union[str, _modes._Mode] = _modes._REPLICA_MODE,
+        comms: Optional[dict[int, _Communicator]] = None,
+    ) -> 'DistributedArray':
+        """Instantiate a distributed array using arguments for its attributes.
+
+        :func:`~cupy.distributed.array.distributed_array` provides a more user
+        friendly way of creating a distributed array from another array.
+
+        Args:
+            shape (tuple of ints): Length of axes.
+            dtype: Data type. It must be an argument of :class:`numpy.dtype`.
+            mode (str or mode object): Mode that determines how overlaps of
+                chunks are interpreted.
```

Comment:
I would like a description of the available models in the docstring under a `.. note:` section :)